spec.bs

```diff
@@ -139,6 +229,85 @@ The following are the [=event handlers=] (and their corresponding [=event handle
       <td><dfn event for="AppHistory">navigateerror</dfn>
 </table>
 
+<div algoithm>
+  To <dfn for="AppHistory">update the entries</dfn> for an {{AppHistory}} instance |appHistory|:
+
+  1. Let |browsingContext| be |appHistory|'s [=relevant global object=]'s [=associated Document=]'s [=Document/browsing context=].
+
+  1. [=Assert=]: |browsingContext| is not null.
+
+  1. Let |sessionHistory| be |browsingContext|'s [=session history=].
+
+  1. Let |currentSHE| be |sessionHistory|'s [=session history/current entry=].
+
+  <!-- TODO make more precise using https://github.com/whatwg/html/pull/4691 -->
+  1. If |sessionHistory|'s [=list/size=] is 1 and |currentSHE|'s [=session history entry/document=] is the initial `about:blank` {{Document}}, then:
+
+    1. Assert: |appHistory|'s [=AppHistory/entry list=] [=list/is empty=].
+
+    1. Return.
+
+    <p class="note">This can occur if running the <a spec="HTML">URL and history update steps</a>, e.g. via {{Document/open(unused1, unused2)|document.open()}}, on the initial `about:blank` {{Document}}. The app history API chooses not to expose the initial `about:blank` {{Document}}, so we bail early.
+
+  1. Let |appHistorySHEs| be a new empty list.
+
+  1. Let |backwardIndex| be the index of |currentSHE| within |sessionHistory|, minus 1.
+
+  1. While |backwardIndex| > 0:
+
+    1. Let |she| be |sessionHistory|[|backwardIndex|].
+
+    1. If |she|'s [=session history entry/origin=] is [=same origin=] with |currentSHE|'s [=session history entry/origin=], then [=list/prepend=] |she| to |appHistorySHEs|.
+
+    1. Otherwise, [=iteration/break=].
+
+    1. Set |backwardIndex| to |backwardIndex| &minus; 1.
+
+  1. Let |forwardIndex| be the index of |currentSHE| within |sessionHistory|.
+
+  1. While |forwardIndex| &lt; |sessionHistory|'s [=list/size=]:
+
+    1. Let |she| be |sessionHistory|[|forwardIndex|].
+
+    1. If |she|'s [=session history entry/origin=] is [=same origin=] with |currentSHE|'s [=session history entry/origin=], then [=list/append=] |she| to |appHistorySHEs|.
+
+    1. Otherwise, [=iteration/break=].
+
+    1. Set |forwardIndex| to |forwardIndex| + 1.
+
+  1. Let |newCurrentIndex| be the index of |currentSHE| within |appHistorySHEs|.
```

Comment:
I don't think `currentShe` has been inserted into `appHistorySHEs` by this point?

Reply:
It was appended during the forward loop, since we set forwardIndex initially to the index of currentSHE within sessionHistoryEntry. Is that too subtle? Maybe there should be a dedicated step for inserting currentSHE?

Reply:
I think either a separate state or a non-normative note might be helpful. Or maybe I'm just biased by my implementation :)

Reply:
Done. It is indeed nice and symmetrical now.